src/GapicClientTrait.php

```diff
@@ -250,12 +264,37 @@ private function buildClientOptions(array $options)
 
         // mTLS: If no apiEndpoint has been supplied by the user, and either
         // GOOGLE_API_USE_MTLS_ENDPOINT tells us to, or mTLS is available, use the mTLS endpoint.
-        if ($options['apiEndpoint'] === $defaultOptions['apiEndpoint']
-            && $this->shouldUseMtlsEndpoint($options)
-        ) {
-            $options['apiEndpoint'] = self::determineMtlsEndpoint($options['apiEndpoint']);
+        if (is_null($apiEndpoint) && $this->shouldUseMtlsEndpoint($options)) {
+            $apiEndpoint = self::determineMtlsEndpoint($defaultOptions['apiEndpoint']);
         }
 
+        $this->apiEndpointFunction = Closure::fromCallable(function () use ($apiEndpoint, $defaultOptions) {
+            // If a custom endpoint is set, or we are using an MTLS endpoint, return it
+            if ($apiEndpoint) {
+                return $apiEndpoint;
+            }
+
+            // If no API endpoint is set, derive the endpoint from the service address template and the
+            // universe domain fetched from the credentials.
+            if (defined('self::SERVICE_ADDRESS_TEMPLATE')) {
```

Comment:
This is TDB but the idea is this will be a new constant on GAPIC clients for constructing the API endpoint from the universe domain.